### PR TITLE
MLPAB-768 - Detect changes between pushes and deploy last successful build

### PR DESCRIPTION
**Pull request by andrewpearce-digital**
# Purpose

Save time for the team by only building when needed

Fixes MLPAB-768

## Approach

- detect changes in app since last push/run of pipeline
- pull tag for current env

truth table for boolean branch_changed_detected and boolean changes_detected

changes_detected = difference to main
branch_changes_detected = difference since last push event

| difference to main | difference since last push event | result                             |
|--------------------|----------------------------------|------------------------------------|
| false              | false                            | deploy production image tag        |
| true               | true                             | build a new image/tag              |
| true               | false                            | deploy last image tag for this env |
| false              | true                             | build new tag (removing files can trigger this)       |

* [x] no app change between main and and commit pulls production tag
* [x] merge app changes from main creates new tag
* [x] push app change commmit creates tag
* [x] push non-app change commit pulls tag
* [x] push app change commit among non-app change commits

## Learning

- https://git-scm.com/docs/git-diff
- https://docs.github.com/en/actions/learn-github-actions/contexts#github-context
- https://docs.github.com/en/webhooks-and-events/webhooks/webhook-events-and-payloads#push

## Checklist

* [x] I have performed a self-review of my own code

### CODE REVIEW

**Comment by andrewpearce-digital** [comment]
attempting to reproduce this functionality on github actions
image

**Comment by andrewpearce-digital** [comment]
Run 1 - No changes to app
image

File: lang/testfile [hunk 1]
```diff
@@ -0,0 +1,2 @@
+This creates a change that should trigger a new build
```
Comment (hawx) [inline lang/testfile:1]: remove these?

Comment (hawx) [reply lang/testfile:1]: or are you wanting to merge and then remove them to test something?

Comment (andrewpearce-digital) [reply lang/testfile:1]: ah, yeah they should be gone
thanks

**Review by acsauk — APPROVED** [review]
:shipit:

**Pull request merged**